src/WidgetMessaging.js

```diff
@@ -109,9 +115,54 @@ export default class WidgetMessaging {
 
     start() {
         this.fromWidget.addEndpoint(this.widgetId, this.widgetUrl);
+        this.fromWidget.addListener("get_openid", this._openIdHandlerRef);
     }
 
     stop() {
         this.fromWidget.removeEndpoint(this.widgetId, this.widgetUrl);
+        this.fromWidget.removeListener("get_openid", this._openIdHandlerRef);
+    }
+
+    async _onOpenIdRequest(ev, rawEv) {
+        if (ev.widgetId !== this.widgetId) return; // not interesting
+
+        const settings = SettingsStore.getValue("widgetOpenIDPermissions");
+        if (settings.blacklist && settings.blacklist.includes(this.widgetId)) {
+            this.fromWidget.sendResponse(rawEv, {state: "blocked"});
+            return;
+        }
+        if (settings.whitelist && settings.whitelist.includes(this.widgetId)) {
```

Comment:
Who controls the `widgetID`? Is there any way that you could whitelist a good widget one day and then later on (maybe in a different room even) an evil widget can use the same widget ID to cheat based on your previous approval of the good one?

Reply:
After chatting more about how this works, I think it's important that we use the same identifiers here as we do for the "allow widget to load" check, which is the tuple of `(roomID, widgetURL)`. It does not seem safe to use `widgetID` for security persistence.